### PR TITLE
Setup cassandra redundancy for GR sites

### PR DESCRIPTION
**Pull request by dtheobald**
This modification sets Cassandra redundancy (of GR systems) based on the network topology gleened from /etc/cassandra/cassandra-topology.properties.

### CODE REVIEW

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh [hunk 1]
```diff
@@ -1,8 +1,119 @@
-#! /bin/bash
-if [[ ! -e /var/lib/cassandra/data/memento ]];
-then
-    echo "CREATE KEYSPACE memento WITH strategy_class='org.apache.cassandra.locator.SimpleStrategy' AND strategy_options:replication_factor=2;
-          USE memento;
-          CREATE TABLE call_lists (impu text PRIMARY KEY, dummy text) WITH read_repair_chance = 1.0;
-    " | cqlsh -2
+#!/bin/bash
+
+keyspace=$(basename $0|sed -e 's#^\(.*\)[.]sh$#\1#')
```
Comment (AlexHockey) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: We know the keyspace is memento - why do this?

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh [hunk 2]
```diff
@@ -1,8 +1,119 @@
...
+    while [ $? -ne 0 ]; do
+        sleep 1
+        printf "${header}."
+        header=""
+        let "cnt=$cnt + 1"
+        if [ $cnt -gt 120 ]; then
+            printf "*** ERROR: Cassandra did not come online!\n"
+            exit 1
+        fi
+        netstat -na | grep -q ":7199[^0-9]"
```
Comment (AlexHockey) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: Missing namespace prefix.

Comment (AlexHockey) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: This is the same command as above. Should be subfunctioned.

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh [hunk 3]
```diff
@@ -1,8 +1,119 @@
...
+    if [ $EUID -ne 0 ]; then
+        echo "When using multiple networks, schema creation must be run as root"
+        exit 2
+    fi
+    namespace_prefix="ip netns exec $signaling_namespace"
+fi
+
+if [[ ! -e /var/lib/cassandra/data/${keyspace} ]]; then
+    header="Waiting for Cassandra"
+    let "cnt=0"
```
Comment (AlexHockey) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: Wouldn't it be easier to just loop 120 times and break if the poll succeeds. Avoids all this nasty bash maths.

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh [hunk 4]
```diff
@@ -1,8 +1,119 @@
...
+
+    $namespace_prefix cqlsh -2 < /tmp/$$.cqlsh.in
+    rm -f /tmp/$$.cqlsh.in
+
+    echo "
+    USE ${keyspace};
+    CREATE TABLE call_lists (impu text PRIMARY KEY, dummy text) WITH read_repair_chance = 1.0;" >> /tmp/$$.cqlsh.in
+
+    $namespace_prefix cqlsh -2 < /tmp/$$.cqlsh.in
+    rm -f /tmp/$$.cqlsh.in
```
Comment (AlexHockey) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: Can't you just build the command and pipe it to cqlsh (like the code that you removed did)?

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh [hunk 5]
```diff
@@ -1,8 +1,119 @@
...
+    while [ $? -ne 0 ]; do
+        sleep 1
+        printf "${header}+"
+        header=""
+        let "cnt=$cnt + 1"
+        if [ $cnt -gt 120 ]; then
+            printf "*** ERROR: Cassandra did not come online!\n"
+            exit 1
+        fi
+        $namespace_prefix netstat -na | grep "LISTEN" | awk '{ print $4 }' | grep -q ":9160\$"
```
Comment (AlexHockey) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: Again, should be subfunctioned.

**Comment by AlexHockey** [comment]
This PR basically seems to break down into two parts. The bit to wait for Cassandra to come online before programming the schema generally looks OK, but I have put some review comments inline. 

I am less happy with the bit to set the replication factor. As far as I can tell, you are trying to do the following:

* Determine if we are in a GR deployment by examining the snitch. 
* If we are GR, set the replication factor in each DC to be either 1 or 2, depending on the number of nodes present in the datacenter.
* Otherwise, just set the replication factor to 1 or 2 depending on the number of nodes in the cluster. 

My main issue is that I don't think the way you are determining the number of nodes is reliable. 

* For the non-GR case you are counting the number of seeds in `cassandra.yaml`. This is a list of existing nodes that the current node can bootstrap from when it joins the cluster. It never contains the full list of nodes in the cluster. In fact, when you add a 2nd node to the cluster it only contains the 1st node, so you will end up setting the replication factor to 1 and making the cluster non-redundant.
* For the GR case you use the entries in `cassandra-topology.properties`. However this supports a `default` option (see http://www.datastax.com/docs/1.1/initialize/cluster_init_multi_dc) so this count isn't accurate either. 

I think it is probably better to just set the replication factor to 2 in each datacenter regardless of the number of nodes. I think Cassandra will be smart enough to cope if you set the factor to 2 but there is only one node in the datacenter/cluster. 

(@bossmc, please could you confirm I've not said anything wrong in the above paragraphs?).

Having said all of that, the way you determine if we are in a GR deployment is not robust. You look at the snitch configured in `cassandra.yaml` using grep and sed. YAML is a complicated format and I'm fairly sure I could write a valid `cassandra.yaml` which broke your parsing. However YAML is machine readable and many scripting languages have built in parsers (python and ruby for example). If you need to read `cassandra.yaml` I think you should be using one of these options. 

In fact using python might tidy up the rest of the script. Python is much better than bash at dealing with structured data and doing maths. It would help avoid some of the more cryptic lines like

```
(for e in ${dcs[@]}; do [[ "$e" == "${dc}" ]] && exit 0; done; exit 1)
```

I'll leave it with you to decide how much you want to migrate to python/ruby and how much to keep in bash. If you do persevere in bash please could you:

* Add a few more comments to your code. I find bash is hard to read (and I don't think I'm alone in this) so err on the side of writing lots of comments explaining what you are trying to achieve. 
* Break your code down into subfunctions, as this will improve readability further.

**Comment by AlexHockey** [comment]
Also, the new code is duplicated across quite a few repos. Could you think about how to commonalize it please?

**Comment by bossmc** [comment]
Some of @AlexHockey's points are less applicable to @ClearwaterCore, where they are more explicit about how to set up Cassandra for GR (no use of `default`, more explicit use of `seeds`).  But for Project Clearwater, we are much laxer and allow users to configure their own Cassandra deployment how they like before connecting the Clearwater components to it.  I'm not sure what that means for this pull request, as it's hard (to impossible) to determine the answers to the questions your script is asking... possibly analysing the `system` keyspace would help?

On @AlexHockey's later comments, I agree that you should be using a proper `YAML` parser, and I'd recommend the one from Python as Python is already installed on Ubuntu images.

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh (hunk 1 above)
Comment (dtheobald) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: The GR code at the bottom of the script is generic and copy-pasted into similar files in other repositories.

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh (hunk 3 above)
Comment (dtheobald) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: Yes, but I am printing out a different progress character depending on what port I'm waiting to come alive.

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh (hunk 4 above)
Comment (dtheobald) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: Yes, but again I was just copy-pasting similar code from elsewhere and not really caring about stuff like that.

**Comment by dtheobald** [comment]
I was just making similar change to all the cassandra schema files. If this change is not appropriate for PC then just reject the pull request and I'll just keep it CC.

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh (hunk 4 above)
Comment (AlexHockey) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: I just grepped the entirety of crest, homestead and memento and didn't find any code like that. It's not how, for example, [homestead_provisioning.sh](https://github.com/Metaswitch/crest/blob/dev/homestead.root/usr/share/clearwater/cassandra-schemas/homestead_provisioning.sh) does things.

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh (hunk 3 above)
Comment (AlexHockey) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: I was suggesting something like the following. I don't care massively though. 

```sh
echo "Waiting for Cassandra"
for ii in {1..120}; do
  $namespace_prefix netstat -na | grep -q ":7199[^0-9]"
  ok=$?
  if [[ $ok != 0 ]]; then
    printf "."
    sleep 1
  fi
done

if [[ $ok != 0 ]]; then
  echo "Cassandra did not come online"
fi

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh (hunk 1 above)
Comment (AlexHockey) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: Saying code can be copy-pasted is not a selling point. 

Seriously though, if this is associated with the copy-pasted code below, why is it miles away from it. Wouldn't it be better to have the copy pasted code to be in a subfunction that I can just call with the name of my keyspace (e.g. `setup_gr_replication "memento"`)

File: memento.root/usr/share/clearwater/cassandra-schemas/memento.sh (hunk 3 above)
Comment (bossmc) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: Though that calls the `netstat` line 120 times even if it passes the first time.  I think you're missing an `else` branch with a `break` statement.

Comment (AlexHockey) [inline memento.root/usr/share/clearwater/cassandra-schemas/memento.sh]: Yeah, but you get my drift. I think this feels much more shell script-y. Like I say I don't care massively though.

**Comment by AlexHockey** [comment]
I would like the GR change if it can be implemented in a way that correctly handles valid `cassandra.yaml` and `cassandra-topology.properties` files. I think this can be achieved as follows:

* Use python to parse `cassandra-topology.properties` to get the snitch.
* If the `ProperyFileSnitch` is in use:
  * Parse the `cassandra-topology.properties` file (coping with a default line) and extract the list of datacenters.
  * Execute an `ALTER KEYSPACE` command to set the replication factor in each datacenter to 2. 
* For other snitches take no action. 

Let me know if you would be willing to implement that.

**Comment by AlexHockey** [comment]
@dtheobald I'm going to close this PR. Like I say, I would like the change if it can be implemented in a way that's more compatible with Project Clearwater. However at this stage it's probably simpler to submit  a new PR.

**Pull request closed without merging**